ipa-core/src/report.rs

```diff
@@ -424,136 +421,384 @@ where
+    pub fn btt_ciphertext(&self) -> &[u8] {
+        &self.data[Self::CIPHERTEXT_BTT_OFFSET..Self::EVENT_TYPE_OFFSET]
+    }
+
+    /// ## Panics
+    /// Only if a `Report` constructor failed to validate the contents properly, which would be a bug.
+    pub fn event_type(&self) -> EventType {
+        EventType::try_from(self.data[Self::EVENT_TYPE_OFFSET]).unwrap() // validated on construction
+    }
+
+    pub fn key_id(&self) -> KeyIdentifier {
+        self.data[Self::KEY_IDENTIFIER_OFFSET]
+    }
+
+    /// ## Panics
+    /// Never.
+    pub fn epoch(&self) -> Epoch {
+        u16::from_le_bytes(
+            self.data[Self::EPOCH_OFFSET..Self::SITE_DOMAIN_OFFSET]
+                .try_into()
+                .unwrap(), // infallible slice-to-array conversion
+        )
+    }
+
+    /// ## Panics
+    /// Only if a `Report` constructor failed to validate the contents properly, which would be a bug.
+    pub fn site_domain(&self) -> &str {
+        std::str::from_utf8(&self.data[Self::SITE_DOMAIN_OFFSET..]).unwrap() // validated on construction
+    }
+
+    /// ## Errors
+    /// If the report contents are invalid.
+    pub fn from_bytes(bytes: B) -> Result<Self, InvalidReportError> {
+        if bytes.len() <= Self::SITE_DOMAIN_OFFSET {
+            return Err(InvalidReportError::Length(
+                bytes.len(),
+                Self::SITE_DOMAIN_OFFSET,
+            ));
+        }
+        EventType::try_from(bytes[Self::EVENT_TYPE_OFFSET])?;
+        let site_domain = &bytes[Self::SITE_DOMAIN_OFFSET..];
+        if !site_domain.is_ascii() {
+            return Err(NonAsciiStringError::from(site_domain).into());
+        }
+        Ok(Self {
+            data: bytes,
+            phantom_data: PhantomData,
+        })
+    }
+
+    /// ## Errors
+    /// If the match key shares in the report cannot be decrypted (e.g. due to a
+    /// failure of the authenticated encryption).
+    /// ## Panics
+    /// Should not panic. Only panics if a `Report` constructor failed to validate the
+    /// contents properly, which would be a bug.
+    pub fn decrypt(
+        &self,
+        key_registry: &KeyRegistry<KeyPair>,
+    ) -> Result<OprfReport<BK, TV, TS>, InvalidReportError> {
+        type CTMKLength = Sum<<Replicated<BA64> as Serializable>::Size, TagSize>;
+        type CTBTTLength<BK, TV, TS> = Sum<
+            Sum<
+                Sum<<Replicated<BK> as Serializable>::Size, <Replicated<TV> as Serializable>::Size>,
+                <Replicated<TS> as Serializable>::Size,
+            >,
+            TagSize,
+        >;
+
+        let info = Info::new(
+            self.key_id(),
+            self.epoch(),
+            self.event_type(),
+            HELPER_ORIGIN,
+            self.site_domain(),
+        )
+        .unwrap(); // validated on construction
+
+        let mut ct_mk: GenericArray<u8, CTMKLength> =
+            *GenericArray::from_slice(self.mk_ciphertext());
+        // let mut ct_mk = self.mk_ciphertext().to_vec();
```

Comment:
don't want to hold this change for too long, but lets make sure we clean up commented code. Fine with a follow up PR to address those